libcnb-test/src/lib.rs

```diff
@@ -0,0 +1,260 @@
+    Other(String),
+}
+
+impl IntegrationTest {
+    /// Creates a new integration test.
+    ///
+    /// # Panics
+    /// - When the connection to Docker failed
+    /// - When the internal Tokio runtime could not be created
+    pub fn new(builder_name: impl Into<String>, app_dir: impl AsRef<Path>) -> Self {
+        IntegrationTest {
+            app_dir: PathBuf::from(app_dir.as_ref()),
+            target_triple: String::from("x86_64-unknown-linux-musl"),
+            builder_name: builder_name.into(),
+            buildpacks: vec![BuildpackReference::Crate],
+            docker: Docker::connect_with_local_defaults()
+                .expect("Could not connect to local Docker deamon"),
+            tokio_runtime: tokio::runtime::Runtime::new()
+                .expect("Could not create internal Tokio runtime"),
+        }
+    }
+
+    /// Sets the buildpacks order.
+    ///
+    /// Defaults to [`BuildpackReference::Crate`].
+    pub fn buildpacks(&mut self, buildpacks: impl Into<Vec<BuildpackReference>>) -> &mut Self {
+        self.buildpacks = buildpacks.into();
+        self
+    }
+
+    /// Sets the target triple.
+    ///
+    /// Defaults to `x86_64-unknown-linux-musl`.
+    pub fn target_triple(&mut self, target_triple: impl Into<String>) -> &mut Self {
+        self.target_triple = target_triple.into();
+        self
+    }
+
+    /// Starts a new integration test run.
+    ///
+    /// This function will copy the application to a temporary directory, cross-compiles this crate,
+    /// packages it as a buildpack and then invokes [pack](https://buildpacks.io/docs/tools/pack/)
+    /// to build a new Docker image with the buildpacks specified by this integration test instance.
+    ///
+    /// Since this function is supposed to only be used in integration tests, failures are not
+    /// signalled via [`Result`](Result) values. Instead, this function panics whenever an unexpected error
+    /// occurred to simplify testing code.
+    ///
+    /// # Panics
+    /// - When the app could not be copied
+    /// - When this crate could not be packed as a buildpack
+    /// - When the "pack" command unexpectedly fails
+    ///
+    /// # Example
+    /// ```no_run
+    /// use libcnb_test::IntegrationTest;
+    ///
+    /// IntegrationTest::new("heroku/buildpacks:20", "test-fixtures/app")
+    ///     .run(|context| {
+    ///         assert!(context.pack_stdout.contains("---> Ruby Buildpack"));
+    ///         assert!(context.pack_stdout.contains("---> Installing bundler"));
+    ///         assert!(context.pack_stdout.contains("---> Installing gems"));
+    ///     })
+    /// ```
+    pub fn run<F: FnOnce(IntegrationTestContext)>(&mut self, f: F) {
+        let temp_app_dir =
+            app::copy_app(&self.app_dir).expect("Could not copy app to temporary location");
+
+        let temp_crate_buildpack_dir =
+            build::package_crate_buildpack(&self.target_triple, temp_app_dir.path())
```

Comment:
Compiling/packaging the buildpack here (vs having a `PackagedBuildpack` type passed into the integration test) does mean we're repeating the compile (which is close to no-op) and packaging each time. However I suppose if the integration tests are run in parallel, this is unavoidable, unless there is a way to run the step prior to any integration tests being run.

Testing locally a `cargo libcnb package` takes ~0.7s, so is not too significant an overhead per test, compared to the rest of the pack/docker/build overhead.

Though I wonder if cargo's locking around the build might slow down the integration tests being run in parallel, if the 10th integration test can't even start until 9x0.7s has passed?

Anyway, likely not something to worry about for now, mainly documenting my initial thoughts.

Reply:
This is a hard problem. I thought about solutions a bit, even to the point of writing a custom test runner. Bottom line, we'd need global state that keeps track of the compilation and packaging state that also works across threads. To simplify the code for this first iteration of `libcnb-test`, I decided to not bother tackling this problem.